### DISCUSSION TITLE
Reflection has strange parsing of `GENERATED ALWAYS AS` expressions for sqlite

### DISCUSSION THREAD

**Discussion by nexushoratio** [sqlalchemy/sqlalchemy#11580] [Usage Questions] [answered]
As a user, my preference is to use software as distributed by my OS vendor (in this case, Debian/stable).  I try not to use things such as `pip`.  So, it is possible, if not even likely, that my issue is fixed with a released version.

It is also very probably that this is simply a user error!

The XY-Problem:  This is just a toy project for learning purposes, and as I'm playing with it, I want to detect changes to the schema and do something appropriate (raise exception, drop table, whatever).  Currently I am doing that by simple string comparisons of the output of `CreateTable()` from the ORM defined tables against the reflected tables, which is how I first noticed the problem below.  If there is a better way of accomplishing this, I'd be interested.  Let me know if I should start a separate thread for that.

Also, I tried to include as much environmental information as I thought was useful in the test case.  But, is there a better way of getting the sqlalchemy version than accessing `sqlalchemy.__version__`?  I searched, but could not find anything.  Also let me know if there was more I should include.

Meanwhile, the reason for this message:

It appears that the computed expressions are not parsed correctly during reflection.  I've noticed at least three distinct issues, though it is likely two are really the same:
1. The closing `)` of an expression is pulled into the *computed.sqltext*.
2. If the expression is a function (e.g., `INSTR()`), it is missing completely.
3. If computed column is persisted, the `STORED` string is listed twice in the resulting `CREATE TABLE` (and *computed.sqltext*).

I think the first and last are really the same issue.  I did not delve completely into the code, but I would not be surprised if the value in a column's *computed.sqltext* is what is used when `str(CreateTable(table))` is called.  In looking at the logs, I see that when sqlalchemy calls `SELECT sql FROM sqlite_master ...;`, it does get back to the correct expression.

Test case:
```python
import os
import platform
import sqlalchemy

engine = sqlalchemy.create_engine(os.getenv('DB_URL'))

DDL = [
    """CREATE TABLE test1 (
        s VARCHAR,
        x VARCHAR GENERATED ALWAYS AS (s || "x")
    );""",
    """CREATE TABLE test2 (
        s VARCHAR,
        x INTEGER GENERATED ALWAYS AS (INSTR(s, ","))
    );""",
    """CREATE TABLE test3 (
        s VARCHAR,
        x VARCHAR GENERATED ALWAYS AS (s || "x") STORED
    );""",
    """CREATE TABLE test4 (
        s VARCHAR,
        x INTEGER GENERATED ALWAYS AS (INSTR(s, ",")) STORED
    );""",
];

conn = engine.connect()

print(platform.freedesktop_os_release()['PRETTY_NAME'])
print(f'{platform.python_version()=}')
print(f'{sqlalchemy.__version__=}')
print(f'{conn.dialect.driver=}')
print(f'{conn.dialect.server_version_info=}')

with conn.begin():
    for ddl in DDL:
        conn.execute(sqlalchemy.sql.text(ddl))

existing = sqlalchemy.schema.MetaData()
existing.reflect(bind=engine)

for table in existing.tables.values():
    print(sqlalchemy.schema.CreateTable(table))
    for col in table.columns:
        if col.computed:
            # The square brackets exist to make it easier to see empty strings
            print(f'[[{col.computed.sqltext=!s}]]')
```

Output:
```
$ env SQLALCHEMY_WARN_20=1 DB_URL=sqlite:///test.db python t.py
Debian GNU/Linux 12 (bookworm)
platform.python_version()='3.11.2'
sqlalchemy.__version__='1.4.46'
conn.dialect.driver='pysqlite'
conn.dialect.server_version_info=(3, 40, 1)

CREATE TABLE test1 (
	s VARCHAR, 
	x VARCHAR GENERATED ALWAYS AS (s || "x")
    ) VIRTUAL
)


[[col.computed.sqltext=s || "x")
    ]]

CREATE TABLE test2 (
	s VARCHAR, 
	x INTEGER GENERATED ALWAYS AS () VIRTUAL
)


[[col.computed.sqltext=]]

CREATE TABLE test3 (
	s VARCHAR, 
	x VARCHAR GENERATED ALWAYS AS (s || "x") STORED
    ) STORED
)


[[col.computed.sqltext=s || "x") STORED
    ]]

CREATE TABLE test4 (
	s VARCHAR, 
	x INTEGER GENERATED ALWAYS AS () STORED
)


[[col.computed.sqltext=]]
```

From `test1`, you can see that the *sqltext* includes the closing paren.  Considering other observations, I was surprised that it was not duplicated in `CREATE TABLE`.

From `test2`, we see the function is missing.

From `test3`, we see that extra `) STORED\n    ` in both *sqltext* and the `CREATE TABLE`.

From `test4`, we see that *sqltext* is empty, and unsurprising, both the function and extra `) STORED` are not present in the `CREATE TABLE`.

In my real code, I did also try `CreateTable(tabled).compile(bind=engine)` to (hopefully) make sure that sqlite specific DDL class was used; it made no difference in either that or this test case.

Thanks for any feedback!

**Comment by CaselIT** [comment, marked as answer]
Hi,

thanks for reporting. created issue #11582

**Reply by nexushoratio** [reply]
Oh, wow.  I'm surprised to find a real issue rather than it being a PEBCAK!
